Cube support

kirkbrauer opened the pull request:
Here's my implementation of the cube type from #698.

The binary send/recv functions are not documented anywhere at the moment, so I had to read the PostgreSQL source files to determine the binary type of the cube type.

https://github.com/postgres/postgres/blob/c30f54ad732ca5c8762bb68bbe0f51de9137dd72/contrib/cube/cube.c

From what I found, there appears to be a 32-bit header that indicates the number of dimensions (i31) in the cube and whether the cube is a point (the first bit) meaning it has the same lower-left and upper-right coordinates. After that there is a linear list of doubles containing all the coordinates in the cube, if the cube is a point they don't bother storing the upper-right coordinates since they are the same as the lower-left ones.

For example, here is a hex dump of the cube `(1)`:
```
00000000  80 00 00 01 3F F0 00 00 00 00 00 00 -> (1)
```

And here is a hex dump for a cube with lower-left and upper-right coordinates:
```
00000000  00 00 00 02 40 14 00 00 00 00 00 00 40 08 00 00
00000010  00 00 00 00 40 00 00 00 00 00 00 00 40 18 00 00
00000020  00 00 00 00 -> (5, 3),(2, 6)
```

We also need to determine which parts of the API need to be implemented on the client side so users can do basic cube manipulation and comparison directly in their program (or if we even need this).

roji commented on src/Npgsql/Internal/TypeHandlers/CubeHandler.cs:
```diff
@@ -0,0 +1,66 @@
+        public override NpgsqlCube Read(NpgsqlReadBuffer buf, int len, FieldDescription? fieldDescription = null)
+        {
+            var header = buf.ReadInt32();
+            var dim = header & DimMask;
```
Is there a limit on the number of dimensions? If not, then in theory this is a potentially large data type, and so maybe not appropriate for NpgsqlSimpleTypeHandler.

kirkbrauer replied:
According to the PostgreSQL source code, there is an arbitrary limit of 100 dimensions. I should probably check that when someone creates a cube on the .NET side of things.

roji commented on src/Npgsql/Internal/TypeHandlers/CubeHandler.cs:
```diff
@@ -0,0 +1,66 @@
+            var header = buf.ReadInt32();
+            var dim = header & DimMask;
+            var point = (header & PointBit) != 0;
+            if (!point) dim *= 2;
```
```suggestion
            if (!point)
                dim *= 2;
```

roji commented on src/Npgsql/Internal/TypeHandlers/CubeHandler.cs:
```diff
@@ -0,0 +1,66 @@
+        public override void Write(NpgsqlCube value, NpgsqlWriteBuffer buf, NpgsqlParameter? parameter)
+        {
+            var header = value.Dimensions;
+            if (value.Point) header |= 1 << 31;
```
```suggestion
            if (value.Point)
                header |= 1 << 31;
```

roji commented on src/Npgsql/TypeMapping/GlobalTypeMapper.cs:
```diff
@@ -634,6 +634,17 @@ void SetupLTreeHandlers()
                 }.Build());
             }
 
+            void SetupCubeHandlers()
```
nit: we can probably just move this into SetupMiscHandlers as there's only one mapping.

roji commented on src/Npgsql/NpgsqlTypes/NpgsqlDbType.cs:
```diff
@@ -548,6 +548,16 @@ public enum NpgsqlDbType
+        /// The PostgreSQL cube type, each value is a multidimensional cube.
+        /// </summary>
+        /// <remarks>See http://www.postgresql.org/docs/current/static/cube.html</remarks>
+        Cube = 70, // Extension type
```
```suggestion
        Cube = 63, // Extension type
```

roji commented on src/Npgsql/NpgsqlTypes/NpgsqlCube.cs:
```diff
@@ -0,0 +1,155 @@
+        /// </summary>
+        /// <param name="s"></param>
+        /// <returns></returns>
+        public static NpgsqlCube Parse(string s)
```
We've had recurring conversations whether to have formatting/parsing on Npgsql-provided types... We definitely want to avoid that in cases where there's a lot of complexity (e.g. date/time types), and the regex above doesn't make me super at-ease either - though this may be simple enough to be OK...

If we do this, we may want to accept a CultureInfo to allow users to specify a culture other than InvariantCulture.

Opinions @vonzshik @Brar @NinoFloris ?

kirkbrauer replied:
Yeah, that regex was terrifying to write and test, but it does work. I'm totally open to suggestions on how to improve this. In the PostgreSQL source they seem to have a custom syntax tree implementation for parsing this, I could make a mirror implementation in pure C# to do this instead.

roji replied:
> users may want to parse a cube that they have stored in a string somewhere else.

But if it's stored somewhere else, why would it necessarily be in the PG format? If it's coming from PG, Npgsql can be used to read it in binary (which is what this PR does), but I'm not sure which scenario would have users dump string representations of cubes to some text files, and then want to read it back...

> Formatting on the other hand is probably necessary so we can maintain parity with how Postgres displays cubes, this will make debugging easier for users who are used to that display format.

So I'm not sure that maintaining parity with PG to this extent is a goal... Let's not forget that even .NET List does not have a string format (i.e. ToString), and in the debugger users should be able to simply drill into the lists directly. Another point is that if users want a string representation of a cube, they can simply request from PG directly (as a string).

To argue against myself, the most convincing argument for supporting formatting (but not necessarily parsing) is for display purposes, e.g. in a GUI.

I know I'm being pretty "conservative" on this point, but we've had some issues with complicated string formatting/parsing on other types, and it's not clear to what extent users actually need it. Another point is that this is also something we can add later upon request.

Opinions @vonzshik @Brar @NinoFloris ?

kirkbrauer replied:
@roji For now, I have decided to remove the parsing/formatting code following your more conservative approach. We can revisit this later if it's requested.

vonzshik replied:
I agree with @roji: since we don't know whether this functionality is actually required, it's better to be conservative for now.

roji commented on src/Npgsql/NpgsqlTypes/NpgsqlCube.cs:
```diff
@@ -0,0 +1,155 @@
+        /// <summary>
+        /// The lower left coordinates of the cube.
+        /// </summary>
+        public IEnumerable<double> LowerLeft => _coords.Take(Dimensions);
```
Using LINQ to split the list into two each time seems a bit inefficient. Would it be more appropriate to just have two separate `List<double>` for LowerLeft/UpperRight, and split/join when reading/writing in the handler? The two could point to the same list instance for point.

kirkbrauer replied:
That's probably the correct approach, I threw this together quickly so I tried to match the way it was done in the PostgreSQL impl. Now that I think about it, that is a much more efficient means to store the data in .NET.

roji commented on src/Npgsql/NpgsqlTypes/NpgsqlCube.cs:
```diff
@@ -0,0 +1,155 @@
+// ReSharper disable once CheckNamespace
+namespace NpgsqlTypes
+{
+    public struct NpgsqlCube : IEquatable<NpgsqlCube>
```
```suggestion
    public readonly struct NpgsqlCube : IEquatable<NpgsqlCube>
```

roji commented on src/Npgsql/NpgsqlTypes/NpgsqlCube.cs:
```diff
@@ -0,0 +1,155 @@
+        public static bool operator !=(NpgsqlCube x, NpgsqlCube y) => !(x == y);
+        
+        public override int GetHashCode() => 
+            Dimensions.GetHashCode() * Coordinates.GetHashCode() * Point.GetHashCode();
```
Use HashCode.Combine.

Also, does Dimensions need to be included given that it's calculated from Point and Coordinates?

kirkbrauer replied:
Probably not, honestly looking at this with fresh eyes, we probably only need to get the hash code of the lower-left and upper-right coordinates and if they point to the same double values, then that will determine if it's a point or not.

roji commented on src/Npgsql/NpgsqlTypes/NpgsqlCube.cs:
```diff
@@ -0,0 +1,155 @@
+        
+        public NpgsqlCube(IEnumerable<double> coords, bool point = false) : this(coords.ToList(), point) { }
+
+        NpgsqlCube(List<double> coords, bool point)
```
Ideally we wouldn't need to expose point as a user-facing boolean flag (even if it's part of the wire representation). See my suggestion below about exposing two separate coordinate lists; if we do that, then we could have a constructor overload that accepts only one list, which would naturally be the point constructor.

kirkbrauer replied:
That's what I have already as a params overload, I will make that change according to your suggestion above to split it into two different lists of coordinates.

roji reviewed (commented):
Thanks, this looks pretty good! See some comments/questions below.

The NpgsqlCube API surface is indeed the thing we need to tie down well. I'm not convinced we need to provide too many client-side operation implementations - the type seems simple enough (basically two coordinate lists) that users can do whatever they want. But am definitely open if you think we should provide some operations.

PS I'm on semi-vacation for the coming week, so expect some delays with responses.

kirkbrauer commented:
@roji What do you think about adding the basic client-side API for manipulating and querying the cube?

I think that the cube type is specific enough that it needs the basic client-side operations so that users can do local math on the cube and manipulate its values. We already have to provide a few of the query methods (cube_is_point, cube_dim, cube_ll_coord, cube_ur_coord) for our implementation to display itself and serialize/deserialize.

My reasoning for this is that manually manipulating the cube's lower-left or upper-right coordinate lists is inherently unsafe and makes it very easy to corrupt the cube. If the number of coordinates in the two lists gets out of sync, we cannot properly serialize/deserialize the cube anymore because such a scenario is not allowed in the cube spec.

Making the `LowerLeft` and `UpperRight` `IEnumerable<double>`s would prevent the user from unsafely changing the dimensions of the cube and require them to use the manipulation API therefore preventing corruption of the cube data. We could also provide utility methods like `AddDimension(double c1, double? c2 = null)` to help with cube manipulation if we want the cube to be a mutable data type.

roji commented:
> Making the LowerLeft and UpperRight IEnumerable<double>s would prevent the user from unsafely changing the dimensions of the cube and require them to use the manipulation API therefore preventing corruption of the cube data

For preventing direct manipulation of the lists, [IReadOnlyList](https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.ireadonlylist-1?view=net-5.0) can be used instead of IEnumerable.

More generally, I'd avoid going in the direction of doing client-side implementations - we generally don't want to fully reproduce PostgreSQL types, with all their complexity. That can easily lead us to a place where we duplicate [the entire PostgreSQL function set](https://www.postgresql.org/docs/current/cube.html) in .NET (because why implement one operation client-side but not another?), and have to maintain it, fix bugs, etc.

We can always start with a minimal implementation, where the data is made fully available, and see what users request; if there's a a lot of need for certain client-side operations, we can always add those later. Otherwise, if you strongly feel we should be providing client-side implementations, do you have certain operations in mind (and reasoning for them)? Or just all of them?

kirkbrauer commented:
That makes sense, I forgot about the specific advantages of the 'IReadOnlyList' interface.

I also agree that we should limit the client-side API tot eh bare minimum operations to start out and then expand it if necessary. I think the best way that users can manipulate the cube is just by using the constructor and the provided lists to re-build a new cube if they desire.

We should probably also leave all the mathematical operations to the database so we don't have to code or debug those.

roji commented:
Great, sounds like we're aligned. There's still an open question in my mind as to whether the type should be mutable or immutable (i.e. if we should expose List or IReadOnlyList). Immutable would allow better enforcing of correctness, but if the lists are long there's a perf hit for having to generate copies etc.

kirkbrauer commented:
@roji The maximum size of each list is 100 elements, so `AddDimension` might be warranted, of course we have to determine if this method returns a new instance of the cube or if it just mutates the existing lists internally.

roji commented:
Yeah. Since the data size here isn't small and fixed, it may make sense to have a fully mutable API and leave it to the user's responsibility to not mess up (e.g. by having different list lengths).

kirkbrauer commented:
That makes sense, if the list sizes get out of sync, we can just pad the opposite array with zeros when serializing or displaying the string rep.

roji commented:
@kirkbrauer just a note to say that this is waiting on you (but absolutely no rush of course!). You can push more commits, and when you'd like another review, ask for one in the reviews panel above.

kirkbrauer commented:
@roji No worries, I got a bit busy last week, but I'll push a new updated version with the changes we discussed above when I get a chance.

kirkbrauer commented:
@roji Quick question regarding the public shipped API vs public unshipped API. Which file would be the best to place the declarations for the NpgsqlCube so we can run the tests? I am not super familiar with these files, but I know that they are commonly used on MS open-source projects to denote the public API. I also know that there isn't great support for Rider, but I'll try to autogenerate them in Visual Studio.

roji commented:
Re the public API files, all changes should be added to the Unshipped file - this is basically a snapshot of all changes done in the current, unreleased version. When we release 6.0, we'll merge those changes into the Shipped file, and the Unshipped file will be empty again.

Unfortunately you're right that the analyzer has issues under Rider - I usually add into Unshipped manually, but VS can indeed be used as well.

kirkbrauer commented on src/Npgsql/NpgsqlTypes/NpgsqlCube.cs:
```diff
@@ -0,0 +1,140 @@
+            UpperRight = upperRight.ToList();
+
+            if (UpperRight.Count != LowerLeft.Count)
+                throw new ArgumentException("Not a valid cube: Different point dimensions in " + ToString(), nameof(upperRight));
```
Since we fill missing coordinates with zero, we probably don't even need to do this check in the constructor, the data will end up normalized when written to the database.

kirkbrauer commented on src/Npgsql/NpgsqlTypes/NpgsqlCube.cs:
```diff
@@ -0,0 +1,140 @@
+        /// Returns the number of dimensions of the cube.
+        /// </summary>
+        /// <returns></returns>
+        public int Dimensions() => Math.Max(LowerLeft.Count, UpperRight.Count);
```
@roji Maybe not, I'm not sure what the best approach is, but I think it might just be easier to fill in missing coordinates with zeros, so therefore the dimensions is just the length of the largest list. I think this makes the most sense since users can manipulate the lists however they want with the exposed mutable lists.

roji replied:
That's certainly possible... I generally lean on the side of requiring users to provide correct data and not correcting it for them (e.g. by padding zeros) - that can help catch bugs in user code.

kirkbrauer replied:
If we do take that approach, where would you suggest throwing the exceptions? We can't throw when adding to a list because that wouldn't be possible, but it is possible to throw an exception when calling `IsPoint()` or `Dimensions()`. We could also only throw exceptions in the constructor and when attempting to write to the database.

roji replied:
> We could also only throw exceptions in the constructor and when attempting to write to the database.

That seems pretty good. Though throwing from IsPoint and Dimensions also makes sense to me, since they can't return sensible results in that invalid state.

kirkbrauer replied:
They _can_ return a sensible result, but maybe they _shouldn't_ as you said. If we converted `IsPoint` and `Dimensions` into properties, then they would always be safe to call even if the data might not be perfect. However, if we keep them as methods, then it would be totally fine to throw an exception even if that inconveniences the user. I could really go either way as long as it produces a good experience.

roji replied:
These do look like they'd be better off as properties, and I don't think there's any particular problem with throwing exceptions from properties.

kirkbrauer replied:
According to Microsoft, we shouldn't throw exceptions on property getters https://docs.microsoft.com/en-us/dotnet/standard/design-guidelines/property. However, maybe we can make an exception (pun intended).

roji replied:
Yeah, that's one of those guidelines that I'm a bit mitigated about :wink: If it's in the user's best interest that the property throw for invalid state, then I think we should do so, as it allows them to detect errors early. That doesn't seem like a good reason to turn the property into a method, which would penalize everyone else using it correctly...

kirkbrauer commented on src/Npgsql/NpgsqlTypes/NpgsqlCube.cs:
```diff
@@ -0,0 +1,116 @@
+            var ret = 266370105;
+            for (var i = 0; i < Dimensions; i++)
+            {
+                ret ^= PGUtil.RotateShift(LowerLeft.ElementAtOrDefault(i).GetHashCode(), ret % PGUtil.BitsInInt);
```
Good, I wasn't sure if that was allowed since the other type don't do that.

vonzshik commented on src/Npgsql/Internal/TypeHandlers/CubeHandler.cs:
```diff
@@ -0,0 +1,83 @@
+            var header = buf.ReadInt32();
+            var dim = header & DimMask;
+            var point = (header & PointBit) != 0;
+            var lowerLeft = new List<double>();
```
```suggestion
            var lowerLeft = new List<double>(dim);
```

vonzshik commented on src/Npgsql/Internal/TypeHandlers/CubeHandler.cs:
```diff
@@ -0,0 +1,83 @@
+            // Cut short execution if we have a point
+            if (point) return new NpgsqlCube(lowerLeft);
+            
+            var upperRight = new List<double>();
```
```suggestion
            var upperRight = new List<double>(dim);
```

roji commented on src/Npgsql/TypeMapping/GlobalTypeMapper.cs:
```diff
@@ -714,6 +713,14 @@ void SetupMiscHandlers()
                     TypeHandlerFactory = new DefaultTypeHandlerFactory<BitStringHandler, BitArray>()
                 }.Build());
+                
+                this.AddMapping(new NpgsqlTypeMappingBuilder
```
Should help with the test failures:

```suggestion
                AddMapping(new NpgsqlTypeMappingBuilder
```

roji commented:
@kirkbrauer can you please squash and rebase this on the latest main branch?

kirkbrauer commented:
@roji Doing it right now.

kirkbrauer commented:
@roji All squashed and rebased. We should be ready for merge, though we might need to do a little work on the tests first.

roji commented:
@kirkbrauer agreed - everything LGTM, but test coverage currently seems to be mainly for the type handler (reading/writing from PG). Some coverage for the APIs on the type would be good.

vonzshik commented on src/Npgsql/NpgsqlTypes/NpgsqlCube.cs:
```diff
@@ -0,0 +1,119 @@
+        /// <exception cref="FormatException">
+        /// Thrown if the lower-left and upper-right lists have different numbers of coordinates.
+        /// </exception>
+        internal static void ValidateCube(List<double> lowerLeft, List<double> upperRight)
```
```suggestion
        static void ValidateCube(List<double> lowerLeft, List<double> upperRight)
```

vonzshik reviewed (changes requested):
Just as @roji suggested, we should have some tests for the cube API. After that. it should be good enough to merge.

kirkbrauer commented:
@roji I'm working on adding additional tests for the cube type, however I have run into a very obvious problem related to our reference optimizations.

If the cube is a point, we set both the lowerLeft and upperRight lists to the same List instance. This is super efficient, but if a user wants to change one list, they will end up manipulating both lists and the source list since all those values point to the same original list instance.

This is probably fine if we treat the cube as immutable, however, if the user is often creating or manipulating their own cubes, it would produce the (possibly) unexpected behavior especially for users who might not be as familiar with pass-by-reference languages.

In other types such as the NpgsqlPath, we are instead re-instantiating the List objects with the same values within the constructor at the penalty of performance, I think this might be a better approach to maintain parity with the rest of the library.

roji commented:
You raise a good point... Here are some options I'm seeing:

1. Duplicate the user-provided list when constructing a point. As you pointed out, this wouldn't be good for perf (note that NpgsqlPath and others were written a long time ago, and shouldn't constrain this type).
2. Make the lists immutable - also bad for perf.
3. Have a separate type for points, which only exposes a single list. This seems a bit heavy/over-complicated to me, and I also worry a bit about confusion between this new point type and PostGIS points.
4. Keep the 2nd list internally null for points, and duplicate it only when accessed through the public property (a sort of "copy-on-write"). The handler would use a separate, internal access that bypasses duplication. This is still non-ideal perf-wise, as duplication happens when reading the property (and not when updating the list), so certain user access patterns would still cause needless duplication.
5. Simply say that UpperRight is null for points. This would make that property nullable (possibly annoying), but would be simple and consistent.
6. Do nothing, and explicitly document that when using the point constructor (or factory), a point remains a point, and so any change in one list is reflected in the other.

Since this type is generally about reading/writing data from PostgreSQL, I think perf should be prioritized. I actually think the current situation isn't necessarily that bad - some users may trip over it, but with good documentation things should be OK. Otherwise options 4 or 5 above also seem reasonable - what do you think?

(@vonzshik @NinoFloris any thoughts here?)

kirkbrauer commented:
I've been thinking about the best way to go about this. I feel that options 2, and 3 are probably off the table because they introduce additional complexity/confusion that would not benefit end users at all. 

Option 6 could work because we already have that implemented, but this would also create the issues where a user would be locked into a specific type of cube, which reduces flexibility and reusability of code making use of the NpgsqlCube type.

As you mentioned, 5  is probably the most reasonable balance between perf and DX, I like the simplicity of setting the UpperRight property to null, however if we make the UpperRight nullable, it could create a situation where a point has to be treated as a special case in any business logic a user would create. If we deem this OK, then we can go ahead with this approach.

If we wanted to go with option 4, it would make much more sense to just make the lists read-only and introduce custom APIs to add/remove coordinates, that way we can preform any optimizations we want including only duplicating the list on write if the lower-left and upper-right values are different. I'm not convinced this is the best approach because it requires custom cube manipulation methods, but at least it allows us to provide the most optimized internal impl and keeps the list lengths the same, preventing us from having to throw exceptions in the properties.

A more radical solution might be to just totally abandon the `LowerLeft` and `UpperRight` lists concept and make the entire cube an enumerable of named value tuples `(double LowerLeft, double UpperRight)` and provide APIs to manipulate the internal `List<(double, double)>`. This would require duplicating the double for each coordinate of a point, however since `ValueTuple`s are just structs on the stack, the performance shouldn't take too much of a hit compared to a list of `decimal`s and we get the advantage of only having to memory manage a single list.

kirkbrauer commented:
@roji Yeah, I had originally preferred option 5, however the more  I have looked at, the more straightforward making it immutable has seemed. It appears that the authors of PostgreSQL's cube type purposely neglected to include (maybe due to technical reasons) any operators or functions to mutate a cube. Just speculating, but the main reason behind this may be the fact that cubes are used as often indexes, so they don't need to be updated unless the value of the type radically changes. For example, changing a color for #FFFFFF `(255, 255, 255)` to #FF0000 `(255, 0, 0)`, which would require mutating 2/3 values in the cube so you might as well just re-create it and save the updated cube.

They also provide additional utility functions that make this type of manipulation easier, which we may wish to include some of if we take the immutable approach:

```
cube_subset ( cube, integer[] ) → cube
cube_union ( cube, cube ) → cube
cube_inter ( cube, cube ) → cube
cube_enlarge ( c cube, r double, n integer ) → cube
```

But, as mentioned before, we may not want to implement all of these as it could lead to reimplementing the entire API.

roji commented:
Immutable sounds OK to me - let's see if @vonzshik has any input here.

Re operations, trivial "data management" stuff like enlarge may make sense, but as above, I'd rather we really avoided implementing any further "advanced" operations. The goal here IMHO really is to allow users to read and write the type, and not duplicate the entire functionality set which PG offers.

vonzshik commented:
IMO, making the `Cube` immutable is for the best - it should be easier to implement and we can always change that at some later point after it's requested with a specific use case in mind.

kirkbrauer commented:
Based on this conversation, I think this is probably the most sensible cube API to implement initially:

```csharp
public class NpgsqlCube {
    public NpgsqlCube(double coord) {}
    public NpgsqlCube(double lowerLeft, double upperRight) {}
    public NpgsqlCube(double[] coords) {}
    public NpgsqlCube(double[] lowerLeft, double[] upperRight) {}
    public NpgsqlCube(NpgsqlCube cube, double coord) {}
    public NpgsqlCube(NpgsqlCube cube, double lowerLeft, double upperRight) {}

    public IReadOnlyList LowerLeft { get; }
    public IReadOnlyList UpperRight { get; }

    public int Dimensions { get; }
    public bool Point { get; }

    public NpgsqlCube Subset(int[] indexes) { }
    public double Distance(NpgsqlCube other) { throw new NotImplementedException(); }
    public NpgsqlCube Union(NpgsqlCube other) { throw new NotImplementedException(); }
    public NpgsqlCube Intersection(NpgsqlCube other) { throw new NotImplementedException(); }
    public NpgsqlCube Enlarge(double radius, int num) { throw new NotImplementedException(); }
}
```

I omitted all the cube operations from the implementation because I don't think it's wise for us to be doing floating-point math or comparisons without relying on the database. If the user wants to enlarge a cube or find the distance, they can easily do that with their own implementation. Subset is implemented because it is just plucking values from specific indexes and does no comparison or math.

roji commented:
@kirkbrauer that looks good overall to me - we can review in the details on the PR. I still don't think we should implement stuff like Distance, Union, Intersection - those go beyond simply accessing values or mutating (via copy).

But otherwise I think you can go ahead and implement. I'd rebase everything as a single commit on the latest main, as this PR has aged somewhat.

kirkbrauer commented:
@roji will do, I forgot about the separate EF Core provider repository. I will update that PR as well to include the operations that we won't implement on this base type. Those should not be included in this PR.

roji commented:
@kirkbrauer yep, exactly. All PG operations should (in theory) have translations at the EF level, but we shouldn't have .NET implementations of them in the .NET cube type itself (which is what this PR will include).

roji commented:
@kirkbrauer any intentions of continuing work on this?

kirkbrauer commented:
@roji I totally forgot about this. I remember starting on this issue because it aligned with the needs of our business. The feature we were thinking of got shelved, but I think we may be revisiting it soon. 

If there is still interest, I can try to push the changes I have sitting on my local branch, it will probably require a lot of changes since this branch is pretty old though.

roji commented:
@kirkbrauer it's really up to you if you want to spend the time on it :) Even if you don't we'll definitely reuse some of the work already done here etc.

keyneom commented:
@kirkbrauer  I'd love to see Cube support

kirkbrauer commented:
@keyneom @roji Ok, I have done some rebasing and pushing to get everything into a single commit on top of the most recent version of Npgsql. 

Have a look at the API that I have implemented for the `NpgsqlCube` type, it differs slightly from my proposal, but retains most of the features in a way that I think is the most usable for most users. I also added tests to ensure that everything is working, they seem to pass locally for me.

keyneom commented:
I don't know enough to opine strongly on this, but in reviewing the historical conversation of this PR and in reviewing the code as it currently stand I think it looks good. The concept of staying close to the interface and contracts provided by postgres itself makes sense to me.

The only question I have is more so along the lines of whether the code as it currently is provides the linq-to-sql generation you'd expect from other types for read/write, etc. but also for cube specific functions and operators like the distance functions, checking for overlaps, etc. I'd be interested in getting some of those added if not. I'd be willing to do the work for adding some of them (at the very least distance would be useful to me) if either of you could explain a bit of how that is handled.

kirkbrauer commented:
@keyneom Those would be implemented as LINQ query extensions as part of the [efcore.pg ](https://github.com/npgsql/efcore.pg) repository, which provides those LINQ-to-SQL conversions that you're looking for. I had a few in the `NpgsqlCube` class, but I removed them to more closely follow the way that other types like `NpgsqlTsVector` work with EF Core.

If you try to run those extension methods on the client they will throw an exception since they aren't implemented client-side, but when run within a LINQ expression tree, they'll be converted to the appropriate SQL and return the correct result.

keyneom commented:
Awesome, yes, I see that makes sense I hadn't been paying attention to the differences between the two repos, this one handles the underlying types and that one interfaces with efcore. Thanks for the explanation! This looks good to me then!

NinoFloris commented:
@kirkbrauer are you still interested in bringing this over the finish line? Maybe there is something we can do to help?

roji commented:
@Jimmacle for EF Core support we can absolutely do this, but I think we need to merge the lower-level support here first.

NinoFloris commented on src/Npgsql/NpgsqlTypes/NpgsqlCube.cs:
```diff
@@ -0,0 +1,288 @@
+        /// <summary>
+        /// Writes the cube in PostgreSQL's text format.
+        /// </summary>
+        public void Write(StringBuilder stringBuilder)
```
Same feedback here, let's leave this out.

NinoFloris commented on src/Npgsql/Internal/ResolverFactories/CubeTypeInfoResolverFactory.cs:
```diff
@@ -0,0 +1,57 @@
+        {
+            mappings.AddStructType<NpgsqlCube>(CubeTypeName,
+                static (options, mapping, _) => mapping.CreateInfo(options, new CubeConverter()),
+                MatchRequirement.DataTypeName);
```
```suggestion
                isDefault: true);
```

NinoFloris commented on src/Npgsql/NpgsqlTypes/NpgsqlCube.cs:
```diff
@@ -0,0 +1,288 @@
+        /// new NpgsqlCube(new[] { 1.0, 2.0 }, new[] { 3.0, 4.0 }).UrCoord(2); // 4.0
+        /// </code>
+        /// </example>
+        public double UrCoord(int n) => _upperRight[n - 1];
```
Idem

NinoFloris commented on src/Npgsql/NpgsqlTypes/NpgsqlCube.cs:
```diff
@@ -0,0 +1,288 @@
+        /// </exception>
+        public NpgsqlCube(IEnumerable<double> lowerLeft, IEnumerable<double> upperRight)
+        {
+            var lowerLeftArray = lowerLeft.ToArray();
```
```suggestion
            var lowerLeftArray = lowerLeft as double[] ?? lowerLeft.ToArray();
```

roji commented on src/Npgsql/NpgsqlTypes/NpgsqlCube.cs:
```diff
@@ -0,0 +1,288 @@
+        /// cube.Subset(3, 2, 1, 1); // '(5,3,1,1),(8,7,6,6)'
+        /// </code>
+        /// </example>
+        public NpgsqlCube Subset(params int[] indexes)
```
Is there a PG-side operation this corresponds to? If so we should consider aligning the name (though .NET conventions are more important).

kirkbrauer commented on src/Npgsql/NpgsqlTypes/NpgsqlCube.cs:
```diff
@@ -0,0 +1,288 @@
+        /// <summary>
+        /// The maximum dimensions of a cube.
+        /// </summary>
+        public const int MaxDimensions = 100;
```
@NinoFloris I think this is a fair point, I guess we can just allow the user to construct technically invalid `NpgsqCub` types with too many dimensions and let PG throw the error itself. This way if the user does indeed recompile PG to have a higher limit for some reason, we won't have issues supporting it.

kirkbrauer replied:
Removed `MaxDimensions` and checks.

kirkbrauer commented on src/Npgsql/PostgresMinimalDatabaseInfo.cs:
```diff
@@ -83,6 +83,7 @@ static PostgresType[] CreateTypes(bool withMultiranges)
         Add(DataTypeNames.PgLsn, oid: 3220, arrayOid: 3221);
         Add(DataTypeNames.Unknown, oid: 705, arrayOid: 0);
         Add(DataTypeNames.Void, oid: 2278, arrayOid: 0);
+        Add(DataTypeNames.Cube, oid: 100000, arrayOid: 100001);
```
We need to figure out a solution to this problem, this was required to make the tests pass, but probably creates many other unintended consequences.

NinoFloris replied:
This shouldn't be necessary. The test suite allows you to install the extension, at which point the type info loading should be able to find it.

roji replied:
We definitely shouldn't have cube in PostgresMinimalDatabaseInfo.

kirkbrauer commented on src/Npgsql/Internal/Postgres/DataTypeNames.cs:
```diff
@@ -51,6 +51,7 @@ static class DataTypeNames
     public static DataTypeName TimeTz => ValidatedName("pg_catalog.timetz");
     public static DataTypeName Inet => ValidatedName("pg_catalog.inet");
     public static DataTypeName Cidr => ValidatedName("pg_catalog.cidr");
+    public static DataTypeName Cube => ValidatedName("pg_catalog.cube");
```
The `cube` extension type can be installed in any schema not just `pg_catalog`, typically it is installed in `public`, but the user may choose to install it somewhere else. Therefore, this qualifier is actually invalid.

kirkbrauer commented on src/Npgsql/NpgsqlTypes/NpgsqlCube.cs:
```diff
@@ -0,0 +1,288 @@
+        /// new NpgsqlCube(new[] { 1.0, 2.0 }, new[] { 3.0, 4.0 }).LlCoord(2); // 2.0
+        /// </code>
+        /// </example>
+        public double LlCoord(int n) => _lowerLeft[n - 1];
```
Removed and replaced with the array accessors.

kirkbrauer commented on src/Npgsql/NpgsqlTypes/NpgsqlCube.cs:
```diff
@@ -0,0 +1,288 @@
+            var upperRightArray = upperRight.ToArray();
+
+            if (lowerLeftArray.Length != upperRightArray.Length)
+                throw new FormatException($"Not a valid cube: Different point dimensions in {lowerLeftArray} and {upperRightArray}.");
```
Now throwing an `ArgumentException` when the lengths do not match.

kirkbrauer commented:
@roji @NinoFloris I discovered a critical issue with the implementation of the cube type. Since this type is technically an extension type, it won't be installed in the `pg_catalog` schema, instead it will usually be installed in `public` or whatever schema the user chooses to install it in.

It appears that there is currently no mechanism to support extension types within the `GlobalTypeMapper`'s CLR type inference unless we have registered them in `PostgresMinimalDatabaseInfo` which is designed to only handle `pg_catalog` types.

We could pretend that the `cube` type is part of the `pg_catalog`, but I don't think that is a long-term solution.

NinoFloris commented:
No problem, it's something to improve in Npgsql. All external plugins have to rely on DataTypeName anyway, so this part of the inference isn't relevant there. And as it really only affects the plugin types that have NpgsqlDbType cases it hasn't really been worth the effort.
